statescript/executor.go

```diff
@@ -206,13 +240,31 @@ func (l Launcher) ExecuteAll(state, action string, ignoreError bool) error {
 			}
 		}
 
-		if ret := execute(filepath.Join(dir, s.Name()), timeout); ret != 0 {
-			// In case of error scripts all should be executed.
-			if ignoreError {
-				log.Errorf("statescript: ignoring error executing '%s': %d", s.Name(), ret)
-			} else {
-				return errors.Errorf("statescript: error executing '%s': %d",
-					s.Name(), ret)
+		iet := time.Now()
+
+	retrylater:
+		for {
+			switch ret := execute(filepath.Join(dir, s.Name()), timeout); ret {
+			case 0:
+				// success. Break the loop and continue
+				break retrylater
+			case exitRetryLater:
+				if time.Since(iet) <= l.getRetryTimeout() {
+					log.Errorf("statescript: %s requested a retry", s.Name())
+					time.Sleep(l.getRetryInterval())
+					continue
+				}
+				return errors.New("statescript: retry time-limit exceeded")
+			default:
+				// In case of error scripts all should be executed.
+				if ignoreError {
+					log.Errorf("statescript: ignoring error executing '%s': %d", s.Name(), ret)
+				} else {
+					return errors.Errorf("statescript: error executing '%s': %d",
+						s.Name(), ret)
+				}
+				break retrylater
```

Comment:
when rewriting this into a switch statement, which is arguably prettier, I will have to use a break/goto, as a break will only break out of the switch. I like the esthetics of this solution. Kosher or not?